### PR TITLE
Adapt olp-sdk-cpp-authetication to use new Network

### PR DESCRIPTION
**Pull request by sergii-vostrikov**
AuthenticationClient now expects Network class to be set to to it to be
able to handle http requests.
SignIn/Out/User/Up result classes reworked to handle their impl classes
more carefully.

Resolves: OLPEDGE-543

Signed-off-by: Sergii Vostrikov <ext-sergii.vostrikov@here.com>

### CODE REVIEW

File: olp-cpp-sdk-authentication/src/SignOutResultImpl.h [hunk 1]
```diff
@@ -29,9 +29,11 @@ namespace olp {
 namespace authentication {
 class SignOutResultImpl : public BaseResult {
  public:
+  SignOutResultImpl() noexcept;
+
   SignOutResultImpl(
       int status, std::string error,
-      std::shared_ptr<rapidjson::Document> json_document = nullptr);
+      std::shared_ptr<rapidjson::Document> json_document = nullptr) noexcept;
   virtual ~SignOutResultImpl();
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/SignOutResultImpl.h]: override?

File: olp-cpp-sdk-authentication/src/SignUpResultImpl.cpp [hunk 2]
```diff
@@ -19,14 +19,25 @@
 
 #include "SignUpResultImpl.h"
 
+#include "Constants.h"
+#include "olp/core/network/HttpStatusCode.h"
+
 using namespace rapidjson;
 
+namespace {
+constexpr auto USER_ID = "userId";
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/SignUpResultImpl.cpp]:
```suggestion
constexpr auto kUserId = "userId";
```

File: olp-cpp-sdk-authentication/src/SignUpResultImpl.cpp [hunk 3]
```diff
@@ -19,14 +19,25 @@
...
+}
+
 namespace olp {
 namespace authentication {
-static const char* USER_ID = "userId";
 
-SignUpResultImpl::SignUpResultImpl(int status, std::string error,
-                                   std::shared_ptr<Document> json_document)
+SignUpResultImpl::SignUpResultImpl() noexcept
+    : SignUpResultImpl(network::HttpStatusCode::ServiceUnavailable,
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/SignUpResultImpl.cpp]: same

File: olp-cpp-sdk-authentication/test/unit/AuthenticationTests.h [hunk 4]
```diff
@@ -371,6 +378,9 @@ class AuthenticationOnlineTest : public AuthenticationBaseTest {
   std::string GetEmail() {
     return TEST_USER_NAME + "-" + GenerateRandomSequence() + "@example.com";
   }
+
+ private:
+  std::shared_ptr<olp::http::Network> network_;
```
Comment (andescu) [inline olp-cpp-sdk-authentication/test/unit/AuthenticationTests.h]: is this used?

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp [hunk 5]
```diff
@@ -178,74 +170,52 @@ class AuthenticationClient::Impl final {
...
       const AuthenticationCredentials& credentials, const std::string& endpoint,
-      const std::shared_ptr<std::vector<unsigned char> >& request_body,
+      const http::NetworkRequest::RequestBodyType& request_body,
       const AuthenticationClient::SignInUserCallback& callback);
 
  private:
   std::string server_url_;
-  std::weak_ptr<ScopedNetwork> network_ptr_;
-  std::mutex network_ptr_lock_;
+  //  std::mutex network_ptr_lock_;
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]: remove dead code

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp [hunk 6]
```diff
@@ -347,8 +323,14 @@ client::CancellationToken AuthenticationClient::Impl::SignInClient(
             break;
         }
       });
-  return client::CancellationToken(
-      [&network, request_id]() { network.Cancel(request_id); });
+  auto request_id = send_outcome.GetRequestId();
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]: it is not given that you always will have an positive outcome.
See https://github.com/heremaps/here-olp-edge-sdk-cpp/pull/96/files#diff-bbc42a63ded02006254f767d3dc2cd3bR72 on how to handle correctly

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp [hunk 7]
```diff
@@ -258,34 +228,40 @@ AuthenticationClient::Impl::Impl(const std::string& authentication_server_url,
...
+      network_settings_() {}
 
 client::CancellationToken AuthenticationClient::Impl::SignInClient(
     const AuthenticationCredentials& credentials,
     const chrono::seconds& expires_in,
     const AuthenticationClient::SignInClientCallback& callback) {
+  if (!network_) {
+    AuthenticationError result({static_cast<int>(http::ErrorCode::IO_ERROR),
+                                "Cannot sign in while offline"});
+    callback(result);
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]: Is this a sync or an async callback?

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp [hunk 8]
```diff
@@ -383,36 +365,45 @@ client::CancellationToken AuthenticationClient::Impl::AcceptTerms(
 
 client::CancellationToken AuthenticationClient::Impl::HandleUserRequest(
     const AuthenticationCredentials& credentials, const std::string& endpoint,
-    const std::shared_ptr<std::vector<unsigned char> >& request_body,
+    const http::NetworkRequest::RequestBodyType& request_body,
     const SignInUserCallback& callback) {
+  if (!network_) {
+    AuthenticationError result({static_cast<int>(http::ErrorCode::IO_ERROR),
+                                "Cannot handle user request while offline"});
+    callback(result);
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]: same here. needs to be async.

File: olp-cpp-sdk-authentication/include/olp/authentication/Settings.h [hunk 9]
```diff
@@ -20,12 +20,18 @@
...
 #include "AuthenticationApi.h"
-#include "NetworkProxySettings.h"
 
 namespace olp {
+namespace http {
+class Network;
+}
+
 namespace authentication {
 static const std::string kHereAccountProductionTokenUrl =
```
[the remark this answers is not in the document]
Comment (sergii-vostrikov) [reply olp-cpp-sdk-authentication/include/olp/authentication/Settings.h]: Can we do that as a separate task, let's say called "url handling cleanup"? The proposed changes are not related to net networking transition.

File: olp-cpp-sdk-authentication/src/SignInResultImpl.cpp [hunk 10]
```diff
@@ -24,13 +24,21 @@
...
+
 namespace olp {
 namespace authentication {
-static const char* TOKEN_TYPE = "tokenType";
-static const char* USER_ID = "userId";
 
-SignInResultImpl::SignInResultImpl(int status, std::string error,
-                                   std::shared_ptr<Document> json_document)
+SignInResultImpl::SignInResultImpl() noexcept
+    : SignInResultImpl(network::HttpStatusCode::ServiceUnavailable,
```
[the remark this answers is not in the document]
Comment (sergii-vostrikov) [reply olp-cpp-sdk-authentication/src/SignInResultImpl.cpp]: Yes, this is the scope of the task for removing old network.
Extension of http namespace is barely the scope of this ticket.

File: olp-cpp-sdk-authentication/include/olp/authentication/Settings.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (andescu) [reply olp-cpp-sdk-authentication/include/olp/authentication/Settings.h]: Sure. Please create a task in backlog

File: olp-cpp-sdk-authentication/test/unit/AuthenticationTests.h (hunk 4 above)
Comment (sergii-vostrikov) [reply olp-cpp-sdk-authentication/test/unit/AuthenticationTests.h]: removed

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp (hunk 6 above)
Comment (sergii-vostrikov) [reply olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]: Do you refer ln107 and on?

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp (hunk 7 above)
Comment (sergii-vostrikov) [reply olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]: same approach as in your reference: https://github.com/heremaps/here-olp-edge-sdk-cpp/pull/96/files#diff-bbc42a63ded02006254f767d3dc2cd3bR72

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp (hunk 8 above)
Comment (sergii-vostrikov) [reply olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]: ditto

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp (hunk 6 above)
Comment (andescu) [reply olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]: Yes, seems line number mismatch. Theoretically you need to check the outcome:

```cpp
if (!send_outcome.IsSuccessfull()) {
    // TODO: Get error and create error result
    callback(result);
  }

   auto request_id = send_outcome.GetRequestId();
```

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp (hunk 7 above)
Comment (andescu) [reply olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]: Except that one is in the context of the async TaskScheduler. I am not sure in which context this is happening.

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp (hunk 6 above)
Comment (sergii-vostrikov) [reply olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]: done

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp (hunk 7 above)
Comment (sergii-vostrikov) [reply olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]: done

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp [hunk 11]
```diff
@@ -258,34 +265,47 @@ AuthenticationClient::Impl::Impl(const std::string& authentication_server_url,
...
     const AuthenticationCredentials& credentials,
     const chrono::seconds& expires_in,
     const AuthenticationClient::SignInClientCallback& callback) {
+  if (!network_) {
+    auto task = [callback] {
+      AuthenticationError result({static_cast<int>(http::ErrorCode::IO_ERROR),
+                                  "Cannot sign in while offline"});
+      callback(result);
+    };
+    if (task_scheduler_) {
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]: should be moved to a anonymous function

```cpp
void ExecuteOrSchedule(std::shared_ptr<thread::TaskScheduler>& task_scheduler,
                       thread::TaskScheduler::CallFuncType&& func) {
  if (!task_scheduler) {
    // User didn't specify a TaskScheduler, execute sync
    func();
    return;
  }

  // Schedule for async execution
  task_scheduler->ScheduleTask(std::move(func));
}
```

Comment (sergii-vostrikov) [reply olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]: done

File: olp-cpp-sdk-authentication/src/AuthenticationClient.cpp [hunk 12]
```diff
@@ -347,72 +375,104 @@ client::CancellationToken AuthenticationClient::Impl::SignInClient(
...
                            generateAcceptTermBody(reacceptanceToken), callback);
 }
 
 client::CancellationToken AuthenticationClient::Impl::HandleUserRequest(
     const AuthenticationCredentials& credentials, const std::string& endpoint,
-    const std::shared_ptr<std::vector<unsigned char> >& request_body,
+    const http::NetworkRequest::RequestBodyType& request_body,
     const SignInUserCallback& callback) {
+  if (!network_) {
+    if (!network_) {
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/AuthenticationClient.cpp]: Copy-paste error?

**Pull request merged**